### DISCUSSION TITLE
Does `AssumeRoleProvider` refresh expired credentials?

### DISCUSSION THREAD

**Discussion by lcmgh** [awslabs/aws-sdk-rust#1055] [Q&A] [answered]
Environment: OIDC authorized Kubernetes POD that sets AWS environment variables to make it work with `from_env`

```rust
let provider = aws_config::sts::AssumeRoleProvider::builder(role_name.clone())
        .region(Region::new(region.clone()))
        .session_name(session_name.unwrap_or("my_service_session".into()))
        .build(
            Arc::new(EnvironmentVariableCredentialsProvider::new_with_env(
                credentials_provider,
            )) as Arc<_>,
        );
 
let config: SdkConfig = aws_config::from_env()
    .credentials_provider(provider)
    .load()
    .await;
 
```

**Comment by ysaito1001** [comment, marked as answer]
Hi @lcmgh, to answer your question, yes credentials will get refreshed. To be more precise, given the provided code snippet above, the `AssumeRoleProvider` will be backed by [LazyCache](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/aws-smithy-runtime/src/client/identity/cache/lazy.rs#L219-L225) whose [resolve_cached_identity](https://github.com/smithy-lang/smithy-rs/blob/main/rust-runtime/aws-smithy-runtime/src/client/identity/cache/lazy.rs#L292-L370) method will internally call `AssumeRoleProvider::provide_credentials` as credentials get close to their expiration times.
